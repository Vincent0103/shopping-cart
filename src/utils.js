const toUrlSafe = (text) => {
  const unsafeCharsRegex = /\\|\/|:|,|'|https?|(\.[^\s]+|\.+)/g;
  const spaceRegex = /\s+/g;
  return text
    .trim()
    .toLowerCase()
    .replaceAll(unsafeCharsRegex, "")
    .replaceAll(spaceRegex, "-");
};

const toTitle = (text) =>
  text
    .trim()
    .split(/-|\s/g)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(" ");

const priceToNumber = (price) => {
  if (typeof price === "string") {
    const regex = /\s*\$/g;
    if (regex.test(price)) price = price.replace(regex, "");
    price = parseFloat(price);
  }
  return price;
};

const verifyDataTypeOfProduct = (productArray) => {
  if (productArray.length !== 2) {
    throw Error(
      `param: ${productArray} contains invalid amount of expected data, expected: [float, float]`,
    );
  }

  const containsNegatives = productArray.some((item) => item < 0);
  if (typeof productArray[0] === "string") {
    throw Error(
      `Make sure ${productArray[0]} is correctly formatted in a float/int type (e.g. 49.99)`,
    );
  } else if (containsNegatives) {
    throw Error(`${productArray} cannot have negative numbers`);
  }
};
const calculateTotal = (products) => {
  // products param data should be like this: [[price, amount], [price, amount], ...]
  products.forEach((product) => {
    verifyDataTypeOfProduct(product);
  });

  return (
    Math.round(
      products.reduce((prev, curr) => curr[0] * curr[1] + prev, 0) * 100,
    ) / 100
  );
};

export { toUrlSafe, toTitle, priceToNumber, calculateTotal };
