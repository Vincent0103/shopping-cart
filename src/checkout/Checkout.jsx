import React, { useContext } from "react";
import { calculateTotal } from "../utils";
import { AppContext } from "../AppContext";
import PropTypes from "prop-types";

const ProductItem = ({ imgSrc, title, price, quantity }) => {
  const P = ({ text }) => <p className="text-xl font-medium">{text}</p>;

  return (
    <>
      <div className="flex h-full items-center justify-center gap-10 justify-self-start">
        <img className="max-h-36" src={imgSrc} alt={`Product: ${title}`} />
        <h3 className="font-medium">{title}</h3>
      </div>
      <P text={`${price}$`} />
      <P text={quantity} />
      <P text={`${price * quantity}$`} />
    </>
  );
};

const GridContainer = ({ customStyling, children }) => (
  <div
    className={`${customStyling} grid w-full grid-cols-[4fr_1fr_1fr_1fr] items-center justify-items-center`}
  >
    {children}
  </div>
);

const Checkout = () => {
  const { cart } = useContext(AppContext);
  console.log(cart);

  return (
    <section className="relative top-20 flex size-full flex-col items-center justify-center gap-20 font-jost">
      <h1 className="mt-20 text-8xl font-extrabold">Your Cart</h1>
      <div
        className="shadow-extraxl-purple h-full w-11/12 rounded-[60px]
      border-[28px] border-accent-400"
      >
        <div className="flex flex-col items-center justify-center p-14 text-3xl font-bold">
          <GridContainer customStyling={"mb-8"}>
            <h2 className="justify-self-start">Products</h2>
            <h2>Price</h2>
            <h2>Quantity</h2>
            <h2>Total</h2>
          </GridContainer>
          <hr className="col-span-4 w-full border-2 border-secondary-200" />
          {cart.map(({ id, imgSrc, title, price, productAmount }) => (
            <React.Fragment key={id}>
              <GridContainer customStyling={"my-6"}>
                <ProductItem
                  imgSrc={imgSrc}
                  title={title}
                  price={price}
                  quantity={productAmount}
                />
              </GridContainer>
              <hr className="col-span-4 w-full border-secondary-400" />
            </React.Fragment>
          ))}
          <p className="mt-20 text-5xl text-accent-200">
            Overall total:{" "}
            {calculateTotal(
              cart.map(({ price, productAmount }) => [price, productAmount]),
            )}
            $
          </p>
        </div>
      </div>
      <button
        type="button"
        className="mb-20 rounded-md bg-accent-500 px-6 py-2 text-4xl font-bold"
      >
        Confirm and Pay
      </button>
    </section>
  );
};

ProductItem.propTypes = {
  imgSrc: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
};

GridContainer.propTypes = {
  customStyling: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
}

export default Checkout;
