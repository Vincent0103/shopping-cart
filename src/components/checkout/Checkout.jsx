import React, { useContext } from "react";
import { calculateTotal } from "../../utils";
import { AppContext } from "../../AppContext";
import PropTypes from "prop-types";

const ProductItem = ({ imgSrc, title, price, quantity, deviceType }) => {
  const P = ({ text }) => <p className="text-xl font-medium">{text}</p>;

  return deviceType === "desktop" ? (
    <>
      <div className="flex size-full items-center justify-center gap-10 justify-self-start">
        <img
          className="max-h-36 w-1/4 object-contain"
          src={imgSrc}
          alt={`Product: ${title}`}
        />
        <h3 className="w-3/4 text-xl font-medium">{title}</h3>
      </div>
      <P text={`${price}$`} />
      <P text={quantity} />
      <P text={`${price * quantity}$`} />
    </>
  ) : (
    <div className="flex items-center gap-4">
      <img
        className="max-h-28 w-2/5 object-contain"
        src={imgSrc}
        alt={`Product: ${title}`}
      />
      <div className="font-regular flex w-3/5 flex-col gap-0.5">
        <h3 className="text-xl leading-[1.15]">{title}</h3>
        <p className="text-sm text-zinc-500">
          {price}$ x {quantity}
        </p>
        <p className="text-xl font-medium text-amber-400">
          <span className="text-lg font-light text-white">Total: </span>
          {price * quantity}$
        </p>
      </div>
    </div>
  );
};

const GridContainer = ({ customStyling, children }) => (
  <div
    className={`${customStyling} grid w-full grid-cols-[4fr_1fr_1fr_1fr] items-center justify-items-center`}
  >
    {children}
  </div>
);

const Checkout = () => {
  const { cart } = useContext(AppContext);

  return (
    <section className="relative top-20 flex size-full flex-col items-center justify-center gap-10 font-jost">
      <h1 className="mt-16 text-8xl font-extrabold max-md:text-5xl">
        Your Cart
      </h1>
      <div
        className="shadow-extraxl-purple h-full w-11/12 rounded-[60px]
      border-[28px] border-accent-400 max-lg:hidden"
      >
        <div className="flex flex-col items-center justify-center p-10 text-3xl font-bold">
          <GridContainer customStyling={"mb-8"}>
            <h2 className="justify-self-start">Products</h2>
            <h2>Price</h2>
            <h2>Quantity</h2>
            <h2>Total</h2>
          </GridContainer>
          <hr className="col-span-4 w-full border-2 border-secondary-200" />
          {cart.map(({ id, imgSrc, title, price, productAmount }) => (
            <React.Fragment key={id}>
              <GridContainer key={id} customStyling={"my-6"}>
                <ProductItem
                  imgSrc={imgSrc}
                  title={title}
                  price={price}
                  quantity={productAmount}
                  deviceType={"desktop"}
                />
              </GridContainer>
              <hr className="col-span-4 w-full border-secondary-400" />
            </React.Fragment>
          ))}
          <p className="mt-20 text-4xl text-accent-200">
            Overall total:{" "}
            {calculateTotal(
              cart.map(({ price, productAmount }) => [price, productAmount]),
            )}
            $
          </p>
        </div>
      </div>
      <div
        className="shadow-extraxl-purple mx-3 flex h-full flex-col
      items-center justify-center rounded-[40px] border-[10px] border-accent-400 lg:hidden"
      >
        <div>
          {cart.map(({ id, imgSrc, title, price, productAmount }) => (
            <div className="m-4 flex flex-col gap-4" key={id}>
              <ProductItem
                imgSrc={imgSrc}
                title={title}
                price={price}
                quantity={productAmount}
                deviceType={"mobile"}
              />
              <hr className="col-span-4 w-full border-secondary-400" />
            </div>
          ))}
        </div>
        <p className="mb-4 text-2xl font-bold">
          Overall total:{" "}
          <span className="text-amber-400">
            {calculateTotal(
              cart.map(({ price, productAmount }) => [price, productAmount]),
            )}
            $
          </span>
        </p>
      </div>
      <button
        type="button"
        className="mb-10 rounded-md bg-accent-500 px-3 py-1 text-xl font-bold"
      >
        Confirm and Pay
      </button>
    </section>
  );
};

ProductItem.propTypes = {
  imgSrc: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
};

GridContainer.propTypes = {
  customStyling: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

export default Checkout;
