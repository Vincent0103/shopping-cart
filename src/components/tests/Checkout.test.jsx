import React from "react";
import { describe, it, vi } from "vitest";
import Checkout from "../checkout/Checkout";
import { render, screen } from "@testing-library/react";
import { calculateTotal } from "../../utils";

vi.mock("../checkout/Checkout.jsx", () => ({
  default: ({ cart }) => (
    <>
      {cart.map(({ id, imgSrc, title, price, productAmount }) => (
        <div data-testid="product-container" key={id}>
          <div>
            <img src={`${imgSrc}`} alt={`Product: ${title}`} />
            <h3>{title}</h3>
          </div>
          <p>{price}$</p>
          <p>{productAmount}</p>
          <p>{price * productAmount}$</p>
        </div>
      ))}
      <p>
        Overall total:{" "}
        {calculateTotal(
          cart.map(({ price, productAmount }) => [price, productAmount]),
        )}
      </p>
    </>
  ),
}));

describe("Checkout", () => {
  it("should show the list of product ordered by the user in the cart", () => {
    const mockedCart = [
      {
        id: 0,
        imgSrc: "src0",
        title: "epikHead",
        price: 56.99,
        productAmount: 1,
      },
      {
        id: 1,
        imgSrc: "src1",
        title: "keyboard",
        price: 12,
        productAmount: 4,
      },
    ];

    const container = render(<Checkout cart={mockedCart} />);
    expect(container).toMatchSnapshot();

    const totalPrice = calculateTotal(
      mockedCart.map(({ price, productAmount }) => [price, productAmount]),
    );
    const regex = new RegExp(`overall total: ${totalPrice}`, "i");
    expect(screen.getByText(regex)).toBeInTheDocument();

    screen.getAllByTestId("product-container").forEach((item) => {
      expect(item).toBeInTheDocument();
    });
  });

  it("should empty the checkout if the cart is empty", () => {
    render(<Checkout cart={[]} />);
    expect(screen.queryByTestId("product-container")).not.toBeInTheDocument();
  });
});
