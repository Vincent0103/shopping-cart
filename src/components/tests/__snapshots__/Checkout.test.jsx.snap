// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkout > should show the list of product ordered by the user in the cart 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="product-container"
      >
        <div>
          <img
            alt="Product: epikHead"
            src="src0"
          />
          <h3>
            epikHead
          </h3>
        </div>
        <p>
          56.99
          $
        </p>
        <p>
          1
        </p>
        <p>
          56.99
          $
        </p>
      </div>
      <div
        data-testid="product-container"
      >
        <div>
          <img
            alt="Product: keyboard"
            src="src1"
          />
          <h3>
            keyboard
          </h3>
        </div>
        <p>
          12
          $
        </p>
        <p>
          4
        </p>
        <p>
          48
          $
        </p>
      </div>
      <p>
        Overall total:
         
        104.99
      </p>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="product-container"
    >
      <div>
        <img
          alt="Product: epikHead"
          src="src0"
        />
        <h3>
          epikHead
        </h3>
      </div>
      <p>
        56.99
        $
      </p>
      <p>
        1
      </p>
      <p>
        56.99
        $
      </p>
    </div>
    <div
      data-testid="product-container"
    >
      <div>
        <img
          alt="Product: keyboard"
          src="src1"
        />
        <h3>
          keyboard
        </h3>
      </div>
      <p>
        12
        $
      </p>
      <p>
        4
      </p>
      <p>
        48
        $
      </p>
    </div>
    <p>
      Overall total:
       
      104.99
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
